File: 53-maximum-subarray/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0

        maxEndingHere = maxSoFar = nums[0]

        for num in nums[1:]:
            maxEndingHere = max(num, maxEndingHere + num)
            maxSoFar = max(maxSoFar, maxEndingHere)

        return maxSoFar

File: 53-maximum-subarray/maximum-subarray.go
func maxSubArray(nums []int) int {
    maxEndingHere := nums[0]
    maxSoFar := nums[0]

    for i := 1; i < len(nums); i++ {
        maxEndingHere = max(nums[i], maxEndingHere+nums[i])
        maxSoFar = max(maxSoFar, maxEndingHere)
    }
    return maxSoFar
}
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}